### ISSUE TITLE
SMTP server incorrectly handles SMTPBadSender from validateFrom

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2158]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2158|
|Type|enhancement|
|Created|2006-10-08 19:41:24Z|

It first tells the client that the address was rejected.  It next tells the client that the address was accepted.

<details><summary>Searchable metadata</summary>

```
trac-id__2158 2158
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1160336484000000 1160336484000000
changetime__1160429975000000 1160429975000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Fixed in validatefrom-2158

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. Just one typo in _ebAuthenticated docstring ('tra nslating').

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Fixed that docstring.  Harsh light of day motivates me to make a couple further exception handling cleanups in this branch, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Okay, cleaned up a bit and added some more tests.  This also fixes response decode error handling now.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_6"></a>|@washort set owner to @exarkun|
|-|-|

Looks good, tests pass, and it fixes the problem. Code looks simpler, which is nice.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [18417]) Merge validatefrom-2158

Author: exarkun
Reviewer: therve, washort
Fixes [#2158](https://github.com/twisted/twisted/issues/2158)

Add tests for handling of various IMessageDelivery implementation behaviors,
including exceptions from validateFrom.  Also add tests for failures from
portal.login and base64 decoding of challenge response strings.  Fix bugs
revealed by all of these tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
